### ISSUE TITLE
blueprint returns associations inconsistently

### ISSUE THREAD

**Issue by gotmikhail** [balderdashy/sails#2850]
We are having an issue where a `many-to-many` association is inconsistently returning the associated objects.

Ex. server response to `GET - /appointment/435`
```json
{
      "users": [],
      "appointmentId": 435,
      "scheduledStartDatetime": "2015-04-30T18:30:00.000Z",
      "scheduledEndDatetime": "2015-04-30T18:45:00.000Z",
      "type": "scheduled",
      "callDuration": null,
      "note": "",
      "createdAt": "2015-03-31T18:25:35.000Z",
      "updatedAt": "2015-03-31T18:25:35.000Z"
}
```

DB:
```
select * from appointment_users__user_appointments where appointment_users = 435;
+-----+-------------------+-------------------+
| id  | appointment_users | user_appointments |
+-----+-------------------+-------------------+
| 748 |               435 |                 8 |
+-----+-------------------+-------------------+
```
User portion of Appointment `Model`
```
        users: {
            collection: 'user',
            via: 'appointments',
            dominant: true
        }
```
Appointment portion of User `Model`
```
        appointments: {
            collection: 'appointment',
            via: 'users',
            dominant: false
        }
```
Repeat the server call, no changes in db or code, haven't restarted the server and haven't updated `node_modules`:
`GET - /appointment/435`
```json
{
  "users": [
    {
      "userId": 8,
      "email": "doctorr3@touchcare.com",
      "lastName": "Tor",
      "firstName": "Doc",
      "middleName": null,
      "gender": "m",
      "birthDate": null,
      "type": "provider",
      "photo": "d41e1f3a-78e9-4b95-b6f7-96f70ad1acc1.jpg",
      "isAvailableAdhoc": true,
      "createdAt": "2015-03-06T01:14:19.000Z",
      "updatedAt": "2015-03-30T21:46:54.000Z",
      "provider": 3
    }
  ],
  "appointmentId": 435,
  "scheduledStartDatetime": "2015-04-30T18:30:00.000Z",
  "scheduledEndDatetime": "2015-04-30T18:45:00.000Z",
  "type": "scheduled",
  "callDuration": null,
  "note": "",
  "createdAt": "2015-03-31T18:25:35.000Z",
  "updatedAt": "2015-03-31T18:25:35.000Z"
}
```
Repeated this call a few times w/ the populated `users`, then it happened again where it returned an empty array.

This is not the only instance, it happens with others as well. Will begin to dig into `sails` and `waterline` code...

Any ideas?

**Comment by gotmikhail** [comment]
Actually, we're seeing patterns, but the patterns are inconsistent. What I mean by that is, right now, every 4th call to `GET /appointment/435` is responding with an empty `Array`. The 3 in between are fully populated.

However, with a different call, it's consistently every other.

**Comment by gotmikhail** [comment]
Just ran an automated test that executed a sequence of 4 calls 100 times. The pattern matched for all 100 sets of 4, every 4th call came back w/ an empty `Array` of `users`.

**Comment by gotmikhail** [comment]
Ok, looks like this might need to be moved to `waterline`.

Tracking it through, it's coming down to the `defaultLimit` in `sails.config.blueprints`. We were overriding that for a few specific `routes`, but not overall.

We have a `mysql` `db` connection, so `sails-mysql` is in use.

Logging the actual queries being executed, we get down into `processChildren` in `sails-mysql/lib/adapter.js` (and `_runChildOperations` in `waterline/lib/waterline/query/finders/operations.js`, we have an `INNER JOIN` being executed on the `User` table and the table `User` <-> `Appointment` association table, `appointment_users__user_appointments `.

```
SELECT `user`.`userId`,`appointment_users__user_appointments`.`appointment_users` AS "___appointment_users" FROM `user` AS `user`  INNER JOIN `appointment_users__user_appointments` ON `appointment_users__user_appointments`.`user_appointments` = `user`.`userId` WHERE `user`.`userId` IN (SELECT `appointment_users__user_appointments`.`user_appointments` FROM `appointment_users__user_appointments` WHERE `appointment_users__user_appointments`.`appointment_users` = 435 )  ORDER BY `user`.`userId` ASC LIMIT 30;
```
(Note that the list of fields was much longer, just stripped out everything but `userId` and `appointment_users` for length.)

The problem is, the search criteria is giving us back all `User` <-> `Appointment` associations for users that are associated w/ the `Appointment` we are looking for. That sounds confusing, and it is :)

`User` 8 has 43 appointments, so we have 43 entries in the `appointment_users__user_appointments` table belonging to `User` 8. The `IN()` portion of the query,
```
SELECT `appointment_users__user_appointments`.`user_appointments` FROM `appointment_users__user_appointments` WHERE `appointment_users__user_appointments`.`appointment_users` = 435
```
Gives us back
```
SELECT `appointment_users__user_appointments`.`user_appointments` FROM `appointment_users__user_appointments` WHERE `appointment_users__user_appointments`.`appointment_users` = 435;
+-------------------+
| user_appointments |
+-------------------+
|                 8 |
+-------------------+
```
and
```
SELECT count(*) FROM `appointment_users__user_appointments` WHERE `appointment_users__user_appointments`.`user_appointments` = 8; +----------+
| count(*) |
+----------+
|       43 |
+----------+
```
So, the `INNER JOIN` is going to give us back 43 rows, all being `User 8` and the single field differing between them being `___appointment_users`.

We apply the `LIMIT 30`, and there we go, we got the 1st 30 `Appointment`s, and left off the one we were looking for, `435`.

If we set the `defaultLimit` to `1000`, just for now, then it works and we get the sole user back every time.

Of note, if we tried to apply the `limit` through the URL, `GET - /appointment/435?limit=1000`, that was not overriding the `defaultLimit`.

I would think that applying the `defautLimit` would mean that if there were more than 30 `Users` belonging to that `Appointment`, then you'd only get 30 of them back.

**Comment by gotmikhail** [comment]
The possible solution I'm seeing is in `sails-mysql`. Opening a ticket there.

**Comment by CWyrtzen** [comment]
Ok, can we close this and it will point people to the waterline open issue?

**Comment by acis** [comment]
any update on this issue? This is very important, as the nested models don't get populated correctly, even with setting the waterline limit config to 8000.

I have a many-to-many relationship between Facility and Sport, with 25000 facility instances. This particular Facility has 15 associated sports. This is the sql log for the blueprint Api call:

The first query, for finding the facility is correct, but the second one, to get the sports is

```
(SELECT 
    `sport`.`name`,
    `sport`.`id`,
    `sport`.`createdAt`,
    `sport`.`updatedAt`,
    `facility_sports__sport_facilities`.`facility_sports` AS '___facility_sports'
FROM
    `sport` AS `sport`
        INNER JOIN
    `facility_sports__sport_facilities` ON `facility_sports__sport_facilities`.`sport_facilities` = `sport`.`id`
WHERE
    `sport`.`id` IN (SELECT 
            `facility_sports__sport_facilities`.`sport_facilities`
        FROM
            `facility_sports__sport_facilities`
        WHERE
            `facility_sports__sport_facilities`.`facility_sports` = 50462)
ORDER BY `sport`.`id` ASC
LIMIT 7000)
```

The Outer query is not filtered on the facility id. Adding the    
```
`facility_sports__sport_facilities`.`facility_sports` = 50462)
```
condition to the outer where as well returns the correct result, 15 entries.

The current query has 9610(!!!!)  rows.

**Comment by sailsbot** [comment]
Thanks for posting, @gotmikhail.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
@gotmikhail @acis Thanks y'all.  Just making sure you're good to go here and that this has been resolved-- I don't remember seeing another issue about it, and I can't find one in search, so I'm assuming it's good to go since I can't replicate it using `sails-mysql` and two models in the same datastore.

But if I'm wrong there, just make a new issue and someone from the core team will look into it asap.

**Comment by gotmikhail** [comment]
@mikermcneil Seems like the pull request was left hanging: https://github.com/balderdashy/waterline-sequel/pull/15

That was quite a while ago, but that's where it was left hanging. It very well may have been resolved, that request just stayed open.

**Comment by mikermcneil** [comment]
@gotmikhail thanks- i'll look at it w/ @particlebanana to be sure
